server/src/main/java/org/candlepin/resteasy/AnnotationLocator.java

```diff
@@ -0,0 +1,141 @@
+/*
+ *  Copyright (c) 2009 - ${YEAR} Red Hat, Inc.
+ *
+ *  This software is licensed to you under the GNU General Public License,
+ *  version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ *  implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ *  FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ *  along with this software; if not, see
+ *  http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ *  Red Hat trademarks are not licensed under GPLv2. No permission is
+ *  granted to use or replicate Red Hat trademarks that are incorporated
+ *  in this software or its documentation.
+ */
+package org.candlepin.resteasy;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Binding;
+import com.google.inject.Injector;
+
+import org.jboss.resteasy.spi.metadata.ResourceBuilder;
+import org.jboss.resteasy.spi.metadata.ResourceClass;
+import org.jboss.resteasy.spi.metadata.ResourceLocator;
+import org.jboss.resteasy.spi.metadata.ResourceMethod;
+import org.jboss.resteasy.util.GetRestful;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+
+/**
+ * Holds a mapping of interface methods to concrete methods.
+ *
+ * Inspired by ResourceLocatorMap.
+ *
+ * This map is populated during servlet initialization and then locked. Various filters can use utility
+ * methods to get annotations on the implementations as well as the interfaces, so that we can use
+ * SecurityHole and other annotations on generated resource implementations.
+ */
+public class AnnotationLocator {
+    private static final Logger log = LoggerFactory.getLogger(AnnotationLocator.class);
+
+    private Map<Method, Method> internalMap;
+    private boolean hasBeenInitialized = false;
+
+    private Injector injector;
+
+    @Inject
+    public AnnotationLocator(Injector injector) {
+        // Maintain the insertion order for nice output in debug statement
+        internalMap = new LinkedHashMap<>();
+        this.injector = injector;
+    }
+
+    @SuppressWarnings("rawtypes")
+    public void init() {
```

Comment:
Is there any chance this can be called multiple times by multiple threads for a given instance? If so, we should probably have the method be synchronized so we don't run the (minute?) risk of two simultaneous calls getting around the lock. If not, disregard this comment.

Reply:
I will move to synchronized from `lock` to `init`; that seems better overall. Another quirk copied from `ResourceLocatorMap`.